sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -185,137 +279,90 @@ object ScalaJSPluginInternal {
           else
             new IRTreePrinter(stdout).printTopLevelTree(vfile.tree)
           stdout.flush()
+
+          logIRCacheStats(streams.value.log)
         }
     )
   }
 
+  /** Collect certain file types from a classpath.
+   *
+   *  @param cp Classpath to collect from
+   *  @param filter Filter for (real) files of interest (not in jars)
+   *  @param collectJar Collect elements from a jar (called for all jars)
+   *  @param collectFile Collect a single file. Params are the file and the
+   *      relative path of the file (to its classpath entry root).
+   *  @return Collected elements attributed with physical files they originated
+   *      from (key: scalaJSSourceFiles).
+   */
+  private def collectFromClasspath[T](cp: Def.Classpath, filter: FileFilter,
+      collectJar: VirtualJarFile => Seq[T],
+      collectFile: (File, String) => T): Attributed[Seq[T]] = {
+
+    val realFiles = Seq.newBuilder[File]
+    val results = Seq.newBuilder[T]
+
+    for (cpEntry <- Attributed.data(cp)) {
+      if (cpEntry.isFile && cpEntry.getName.endsWith(".jar")) {
+        realFiles += cpEntry
+        val vf = new FileVirtualBinaryFile(cpEntry) with VirtualJarFile
+        results ++= collectJar(vf)
+      } else if (cpEntry.isDirectory) {
+        for {
+          (file, relPath) <- Path.selectSubpaths(cpEntry, filter)
+        } {
+          realFiles += file
+          results += collectFile(file, relPath)
+        }
+      } else {
+        sys.error("Illegal classpath entry: " + cpEntry.getPath)
+      }
+    }
+
+    Attributed.blank(results.result()).put(scalaJSSourceFiles, realFiles.result())
+  }
+
   val scalaJSConfigSettings: Seq[Setting[_]] = Seq(
       incOptions ~= scalaJSPatchIncOptions
   ) ++ (
-      scalajspSettings
+      scalajspSettings ++
+      stageKeys.flatMap((scalaJSStageSettings _).tupled) ++
+      dispatchTaskKeySettings(scalaJSLinkedFile) ++
+      dispatchSettingKeySettings(scalaJSLinker)
   ) ++ Seq(
+      /* Note: This cache only gets freed by its finalizer. Otherwise we'd need
+       * to intercept reloads in sbt (see #2171).
+       * Also note that it doesn't get cleared by the sbt's clean task.
+       */
+      scalaJSIRCacheHolder := globalIRCache.newCache,
+      scalaJSIRCache <<=
+        Def.task(scalaJSIRCacheHolder.value).dependsOn(scalaJSClearCacheStats),
 
-      scalaJSPreLinkClasspath := {
-        val cp = fullClasspath.value
-        val pcp = PartialClasspathBuilder.build(Attributed.data(cp).toList)
-        val ccp = pcp.resolve(jsDependencyFilter.value, jsManifestFilter.value)
-
-        if (checkScalaJSSemantics.value)
-          ccp.checkCompliance(scalaJSSemantics.value)
```

Comment:
This feature disappeared. The new version of `ScalaJSPluginInternal` does not even mention `checkScalaJSSemantics` other than to set it to `true`.